{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fatty-royal",
   "metadata": {},
   "outputs": [],
   "source": [
    "import swan_vis as swan\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "broken-management",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from /Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/swan/swan.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('/Users/fairliereese/mortazavi_lab/data/rnawg/lr_bulk/swan/swan.p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "concrete-tsunami",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gm12878_1_1</th>\n",
       "      <td>gm12878_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm12878_1_2</th>\n",
       "      <td>gm12878_1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm12878_1_3</th>\n",
       "      <td>gm12878_1_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm12878_1_4</th>\n",
       "      <td>gm12878_1_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm12878_3_1</th>\n",
       "      <td>gm12878_3_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 dataset\n",
       "gm12878_1_1  gm12878_1_1\n",
       "gm12878_1_2  gm12878_1_2\n",
       "gm12878_1_3  gm12878_1_3\n",
       "gm12878_1_4  gm12878_1_4\n",
       "gm12878_3_1  gm12878_3_1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "devoted-arrest",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODEHT000206785</th>\n",
       "      <td>ENCODEHT000206785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHT000206831</th>\n",
       "      <td>ENCODEHT000206831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHT000206867</th>\n",
       "      <td>ENCODEHT000206867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHT000206868</th>\n",
       "      <td>ENCODEHT000206868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEHT000206870</th>\n",
       "      <td>ENCODEHT000206870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 tid\n",
       "tid                                 \n",
       "ENCODEHT000206785  ENCODEHT000206785\n",
       "ENCODEHT000206831  ENCODEHT000206831\n",
       "ENCODEHT000206867  ENCODEHT000206867\n",
       "ENCODEHT000206868  ENCODEHT000206868\n",
       "ENCODEHT000206870  ENCODEHT000206870"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg.adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "anonymous-factory",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/Users/fairliereese/mortazavi_lab/data/mousewg/hippocampus/lr_splitseq/scanpy/transcript_processed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cardiovascular-interstate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc</th>\n",
       "      <th>bc3_long</th>\n",
       "      <th>bc2_long</th>\n",
       "      <th>bc1_long</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sr_umi_count</th>\n",
       "      <th>sr_gene_count</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>sr_celltype</th>\n",
       "      <th>...</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>leiden</th>\n",
       "      <th>lr_celltype</th>\n",
       "      <th>umap_x</th>\n",
       "      <th>umap_y</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bc_index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACCGAGAATTCATGG-ont_a</th>\n",
       "      <td>AAACATCGAACCGAGAATTCATGG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACCGAGA</td>\n",
       "      <td>ATTCATGG</td>\n",
       "      <td>ont_a</td>\n",
       "      <td>41.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>665.0</td>\n",
       "      <td>AAACATCGAACCGAGA_8</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>...</td>\n",
       "      <td>70</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L6</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>12.715929</td>\n",
       "      <td>-1.627943</td>\n",
       "      <td>70</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACCGAGAATTCATGG-pb_a</th>\n",
       "      <td>AAACATCGAACCGAGAATTCATGG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACCGAGA</td>\n",
       "      <td>ATTCATGG</td>\n",
       "      <td>pb_a</td>\n",
       "      <td>41.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>665.0</td>\n",
       "      <td>AAACATCGAACCGAGA_8</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L2</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>10.345781</td>\n",
       "      <td>1.028104</td>\n",
       "      <td>13</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACGCTTATACTCGAA-ont_b</th>\n",
       "      <td>AAACATCGAACGCTTATACTCGAA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACGCTTA</td>\n",
       "      <td>TACTCGAA</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L17</td>\n",
       "      <td>Vascular + endothelial</td>\n",
       "      <td>10.930988</td>\n",
       "      <td>13.045471</td>\n",
       "      <td>14</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACGTGATACTATATA-ont_a</th>\n",
       "      <td>AAACATCGAACGTGATACTATATA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACGTGAT</td>\n",
       "      <td>ACTATATA</td>\n",
       "      <td>ont_a</td>\n",
       "      <td>48.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L1</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>-5.870996</td>\n",
       "      <td>-1.726015</td>\n",
       "      <td>20</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGACAAGCTAGAGCACAA-ont_b</th>\n",
       "      <td>AAACATCGACAAGCTAGAGCACAA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>ACAAGCTA</td>\n",
       "      <td>GAGCACAA</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>90.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L1</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>-5.305627</td>\n",
       "      <td>-1.920617</td>\n",
       "      <td>51</td>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      bc  bc3_long  bc2_long  \\\n",
       "bc_index                                                                       \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  AAACATCGAACCGAGAATTCATGG  AAACATCG  AACCGAGA   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   AAACATCGAACCGAGAATTCATGG  AAACATCG  AACCGAGA   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  AAACATCGAACGCTTATACTCGAA  AAACATCG  AACGCTTA   \n",
       "AAACATCGAACGTGATACTATATA-ont_a  AAACATCGAACGTGATACTATATA  AAACATCG  AACGTGAT   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  AAACATCGACAAGCTAGAGCACAA  AAACATCG  ACAAGCTA   \n",
       "\n",
       "                                bc1_long experiment  sr_umi_count  \\\n",
       "bc_index                                                            \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  ATTCATGG      ont_a          41.0   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   ATTCATGG       pb_a          41.0   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  TACTCGAA      ont_b           NaN   \n",
       "AAACATCGAACGTGATACTATATA-ont_a  ACTATATA      ont_a          48.0   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  GAGCACAA      ont_b           NaN   \n",
       "\n",
       "                                sr_gene_count  Unnamed: 0        cell_barcode  \\\n",
       "bc_index                                                                        \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a           39.0       665.0  AAACATCGAACCGAGA_8   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a            39.0       665.0  AAACATCGAACCGAGA_8   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b            NaN         NaN                 NaN   \n",
       "AAACATCGAACGTGATACTATATA-ont_a           46.0         NaN                 NaN   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b            NaN         NaN                 NaN   \n",
       "\n",
       "                                sr_celltype  ... n_genes_by_counts  \\\n",
       "bc_index                                     ...                     \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  Exc. Neuron  ...                70   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   Exc. Neuron  ...                13   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b          NaN  ...                14   \n",
       "AAACATCGAACGTGATACTATATA-ont_a          NaN  ...                20   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b          NaN  ...                51   \n",
       "\n",
       "                                total_counts  total_counts_mt pct_counts_mt  \\\n",
       "bc_index                                                                      \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a         211.0              0.0           0.0   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a           33.0              0.0           0.0   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b          21.0              0.0           0.0   \n",
       "AAACATCGAACGTGATACTATATA-ont_a          34.0              0.0           0.0   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b          90.0              0.0           0.0   \n",
       "\n",
       "                               leiden             lr_celltype     umap_x  \\\n",
       "bc_index                                                                   \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a     L6             Exc. Neuron  12.715929   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a      L2             Exc. Neuron  10.345781   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b    L17  Vascular + endothelial  10.930988   \n",
       "AAACATCGAACGTGATACTATATA-ont_a     L1               Astrocyte  -5.870996   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b     L1               Astrocyte  -5.305627   \n",
       "\n",
       "                                   umap_y n_genes n_counts  \n",
       "bc_index                                                    \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  -1.627943      70    211.0  \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a    1.028104      13     33.0  \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  13.045471      14     21.0  \n",
       "AAACATCGAACGTGATACTATATA-ont_a  -1.726015      20     34.0  \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  -1.920617      51     90.0  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "accurate-raise",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000142472</th>\n",
       "      <td>ENCODEMT000142472</td>\n",
       "      <td>ENSMUSG00000020671.9</td>\n",
       "      <td>Rab10</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000144512</th>\n",
       "      <td>ENCODEMT000144512</td>\n",
       "      <td>ENSMUSG00000040374.13</td>\n",
       "      <td>Pex2</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000144928</th>\n",
       "      <td>ENCODEMT000144928</td>\n",
       "      <td>ENSMUSG00000096054.3</td>\n",
       "      <td>Syne1</td>\n",
       "      <td>NNC</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>99.966937</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000145365</th>\n",
       "      <td>ENCODEMT000145365</td>\n",
       "      <td>ENSMUSG00000019772.5</td>\n",
       "      <td>Vip</td>\n",
       "      <td>NNC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000145637</th>\n",
       "      <td>ENCODEMT000145637</td>\n",
       "      <td>ENSMUSG00000059734.7</td>\n",
       "      <td>Ndufs8</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      annot_transcript_id          annot_gene_id  \\\n",
       "annot_transcript_name                                              \n",
       "ENCODEMT000142472       ENCODEMT000142472   ENSMUSG00000020671.9   \n",
       "ENCODEMT000144512       ENCODEMT000144512  ENSMUSG00000040374.13   \n",
       "ENCODEMT000144928       ENCODEMT000144928   ENSMUSG00000096054.3   \n",
       "ENCODEMT000145365       ENCODEMT000145365   ENSMUSG00000019772.5   \n",
       "ENCODEMT000145637       ENCODEMT000145637   ENSMUSG00000059734.7   \n",
       "\n",
       "                      annot_gene_name transcript_novelty     mt  \\\n",
       "annot_transcript_name                                             \n",
       "ENCODEMT000142472               Rab10                NIC  False   \n",
       "ENCODEMT000144512                Pex2                NIC  False   \n",
       "ENCODEMT000144928               Syne1                NNC  False   \n",
       "ENCODEMT000145365                 Vip                NNC  False   \n",
       "ENCODEMT000145637              Ndufs8                NIC  False   \n",
       "\n",
       "                       n_cells_by_counts  mean_counts  pct_dropout_by_counts  \\\n",
       "annot_transcript_name                                                          \n",
       "ENCODEMT000142472                      1     0.000165              99.983468   \n",
       "ENCODEMT000144512                      1     0.000331              99.983468   \n",
       "ENCODEMT000144928                      2     0.000331              99.966937   \n",
       "ENCODEMT000145365                      1     0.000496              99.983468   \n",
       "ENCODEMT000145637                      1     0.000165              99.983468   \n",
       "\n",
       "                       total_counts  \n",
       "annot_transcript_name                \n",
       "ENCODEMT000142472               1.0  \n",
       "ENCODEMT000144512               2.0  \n",
       "ENCODEMT000144928               2.0  \n",
       "ENCODEMT000145365               3.0  \n",
       "ENCODEMT000145637               1.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "civilian-frame",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.index.name = 'dataset'\n",
    "adata.var.reset_index(inplace=True)\n",
    "adata.var.set_index('annot_transcript_id', inplace=True)\n",
    "adata.var.index.name = 'tid'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "applicable-spirituality",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000142472</th>\n",
       "      <td>ENCODEMT000142472</td>\n",
       "      <td>ENSMUSG00000020671.9</td>\n",
       "      <td>Rab10</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000144512</th>\n",
       "      <td>ENCODEMT000144512</td>\n",
       "      <td>ENSMUSG00000040374.13</td>\n",
       "      <td>Pex2</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000144928</th>\n",
       "      <td>ENCODEMT000144928</td>\n",
       "      <td>ENSMUSG00000096054.3</td>\n",
       "      <td>Syne1</td>\n",
       "      <td>NNC</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>99.966937</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000145365</th>\n",
       "      <td>ENCODEMT000145365</td>\n",
       "      <td>ENSMUSG00000019772.5</td>\n",
       "      <td>Vip</td>\n",
       "      <td>NNC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000145637</th>\n",
       "      <td>ENCODEMT000145637</td>\n",
       "      <td>ENSMUSG00000059734.7</td>\n",
       "      <td>Ndufs8</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>99.983468</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  annot_transcript_name          annot_gene_id  \\\n",
       "tid                                                              \n",
       "ENCODEMT000142472     ENCODEMT000142472   ENSMUSG00000020671.9   \n",
       "ENCODEMT000144512     ENCODEMT000144512  ENSMUSG00000040374.13   \n",
       "ENCODEMT000144928     ENCODEMT000144928   ENSMUSG00000096054.3   \n",
       "ENCODEMT000145365     ENCODEMT000145365   ENSMUSG00000019772.5   \n",
       "ENCODEMT000145637     ENCODEMT000145637   ENSMUSG00000059734.7   \n",
       "\n",
       "                  annot_gene_name transcript_novelty     mt  \\\n",
       "tid                                                           \n",
       "ENCODEMT000142472           Rab10                NIC  False   \n",
       "ENCODEMT000144512            Pex2                NIC  False   \n",
       "ENCODEMT000144928           Syne1                NNC  False   \n",
       "ENCODEMT000145365             Vip                NNC  False   \n",
       "ENCODEMT000145637          Ndufs8                NIC  False   \n",
       "\n",
       "                   n_cells_by_counts  mean_counts  pct_dropout_by_counts  \\\n",
       "tid                                                                        \n",
       "ENCODEMT000142472                  1     0.000165              99.983468   \n",
       "ENCODEMT000144512                  1     0.000331              99.983468   \n",
       "ENCODEMT000144928                  2     0.000331              99.966937   \n",
       "ENCODEMT000145365                  1     0.000496              99.983468   \n",
       "ENCODEMT000145637                  1     0.000165              99.983468   \n",
       "\n",
       "                   total_counts  \n",
       "tid                              \n",
       "ENCODEMT000142472           1.0  \n",
       "ENCODEMT000144512           2.0  \n",
       "ENCODEMT000144928           2.0  \n",
       "ENCODEMT000145365           3.0  \n",
       "ENCODEMT000145637           1.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "composite-angola",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc</th>\n",
       "      <th>bc3_long</th>\n",
       "      <th>bc2_long</th>\n",
       "      <th>bc1_long</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sr_umi_count</th>\n",
       "      <th>sr_gene_count</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>sr_celltype</th>\n",
       "      <th>...</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>leiden</th>\n",
       "      <th>lr_celltype</th>\n",
       "      <th>umap_x</th>\n",
       "      <th>umap_y</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACCGAGAATTCATGG-ont_a</th>\n",
       "      <td>AAACATCGAACCGAGAATTCATGG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACCGAGA</td>\n",
       "      <td>ATTCATGG</td>\n",
       "      <td>ont_a</td>\n",
       "      <td>41.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>665.0</td>\n",
       "      <td>AAACATCGAACCGAGA_8</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>...</td>\n",
       "      <td>70</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L6</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>12.715929</td>\n",
       "      <td>-1.627943</td>\n",
       "      <td>70</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACCGAGAATTCATGG-pb_a</th>\n",
       "      <td>AAACATCGAACCGAGAATTCATGG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACCGAGA</td>\n",
       "      <td>ATTCATGG</td>\n",
       "      <td>pb_a</td>\n",
       "      <td>41.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>665.0</td>\n",
       "      <td>AAACATCGAACCGAGA_8</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L2</td>\n",
       "      <td>Exc. Neuron</td>\n",
       "      <td>10.345781</td>\n",
       "      <td>1.028104</td>\n",
       "      <td>13</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACGCTTATACTCGAA-ont_b</th>\n",
       "      <td>AAACATCGAACGCTTATACTCGAA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACGCTTA</td>\n",
       "      <td>TACTCGAA</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L17</td>\n",
       "      <td>Vascular + endothelial</td>\n",
       "      <td>10.930988</td>\n",
       "      <td>13.045471</td>\n",
       "      <td>14</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACGTGATACTATATA-ont_a</th>\n",
       "      <td>AAACATCGAACGTGATACTATATA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACGTGAT</td>\n",
       "      <td>ACTATATA</td>\n",
       "      <td>ont_a</td>\n",
       "      <td>48.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L1</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>-5.870996</td>\n",
       "      <td>-1.726015</td>\n",
       "      <td>20</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGACAAGCTAGAGCACAA-ont_b</th>\n",
       "      <td>AAACATCGACAAGCTAGAGCACAA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>ACAAGCTA</td>\n",
       "      <td>GAGCACAA</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>90.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>L1</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>-5.305627</td>\n",
       "      <td>-1.920617</td>\n",
       "      <td>51</td>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      bc  bc3_long  bc2_long  \\\n",
       "dataset                                                                        \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  AAACATCGAACCGAGAATTCATGG  AAACATCG  AACCGAGA   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   AAACATCGAACCGAGAATTCATGG  AAACATCG  AACCGAGA   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  AAACATCGAACGCTTATACTCGAA  AAACATCG  AACGCTTA   \n",
       "AAACATCGAACGTGATACTATATA-ont_a  AAACATCGAACGTGATACTATATA  AAACATCG  AACGTGAT   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  AAACATCGACAAGCTAGAGCACAA  AAACATCG  ACAAGCTA   \n",
       "\n",
       "                                bc1_long experiment  sr_umi_count  \\\n",
       "dataset                                                             \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  ATTCATGG      ont_a          41.0   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   ATTCATGG       pb_a          41.0   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  TACTCGAA      ont_b           NaN   \n",
       "AAACATCGAACGTGATACTATATA-ont_a  ACTATATA      ont_a          48.0   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  GAGCACAA      ont_b           NaN   \n",
       "\n",
       "                                sr_gene_count  Unnamed: 0        cell_barcode  \\\n",
       "dataset                                                                         \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a           39.0       665.0  AAACATCGAACCGAGA_8   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a            39.0       665.0  AAACATCGAACCGAGA_8   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b            NaN         NaN                 NaN   \n",
       "AAACATCGAACGTGATACTATATA-ont_a           46.0         NaN                 NaN   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b            NaN         NaN                 NaN   \n",
       "\n",
       "                                sr_celltype  ... n_genes_by_counts  \\\n",
       "dataset                                      ...                     \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  Exc. Neuron  ...                70   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a   Exc. Neuron  ...                13   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b          NaN  ...                14   \n",
       "AAACATCGAACGTGATACTATATA-ont_a          NaN  ...                20   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b          NaN  ...                51   \n",
       "\n",
       "                                total_counts  total_counts_mt pct_counts_mt  \\\n",
       "dataset                                                                       \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a         211.0              0.0           0.0   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a           33.0              0.0           0.0   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b          21.0              0.0           0.0   \n",
       "AAACATCGAACGTGATACTATATA-ont_a          34.0              0.0           0.0   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b          90.0              0.0           0.0   \n",
       "\n",
       "                               leiden             lr_celltype     umap_x  \\\n",
       "dataset                                                                    \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a     L6             Exc. Neuron  12.715929   \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a      L2             Exc. Neuron  10.345781   \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b    L17  Vascular + endothelial  10.930988   \n",
       "AAACATCGAACGTGATACTATATA-ont_a     L1               Astrocyte  -5.870996   \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b     L1               Astrocyte  -5.305627   \n",
       "\n",
       "                                   umap_y n_genes n_counts  \n",
       "dataset                                                     \n",
       "AAACATCGAACCGAGAATTCATGG-ont_a  -1.627943      70    211.0  \n",
       "AAACATCGAACCGAGAATTCATGG-pb_a    1.028104      13     33.0  \n",
       "AAACATCGAACGCTTATACTCGAA-ont_b  13.045471      14     21.0  \n",
       "AAACATCGAACGTGATACTATATA-ont_a  -1.726015      20     34.0  \n",
       "AAACATCGACAAGCTAGAGCACAA-ont_b  -1.920617      51     90.0  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "missing-obligation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding annotation to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts: 100%|█████████▉| 141800/141862 [01:12<00:00, 1952.16it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 754000/754537 [00:00<00:00, 977718.93it/s] \n",
      "Reindexing edges:  97%|█████████▋| 730000/754537 [00:00<00:00, 806259.21it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Annotation added to the SwanGraph\n",
      "\n",
      "Adding transcriptome to the SwanGraph\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing transcripts: 100%|█████████▉| 153400/153419 [01:10<00:00, 2161.39it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reindexing and sorting entries on genomic location...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reindexing vertices: 100%|█████████▉| 762000/762130 [00:00<00:00, 993769.98it/s] \n",
      "Reindexing edges:  98%|█████████▊| 745000/762130 [00:01<00:00, 738980.93it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Transcriptome added to the SwanGraph\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'SwanGraph' object has no attribute 'write'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-d646b3301016>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_transcriptome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgtf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'test_mousewg.p'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'SwanGraph' object has no attribute 'write'"
     ]
    }
   ],
   "source": [
    "sg = swan.SwanGraph()\n",
    "\n",
    "annot = '/Users/fairliereese/mortazavi_lab/data/mousewg/refs/gencode.vM21.primary_assembly.annotation_UCSC_names.gtf'\n",
    "gtf = '/Users/fairliereese/mortazavi_lab/data/mousewg/lr_bulk/talon/mouse_all_known_nic_nnc_talon.gtf'\n",
    "\n",
    "sg.add_annotation(annot, verbose=True)\n",
    "sg.add_transcriptome(gtf, verbose=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "lonely-reconstruction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving graph as test_mousewg.p\n"
     ]
    }
   ],
   "source": [
    "sg.save_graph('test_mousewg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "african-florence",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/Users/fairliereese/mortazavi_lab/data/mousewg/adrenal/lr_splitseq/scanpy/transcript_processed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "expressed-carolina",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000142671</th>\n",
       "      <td>ENCODEMT000142671</td>\n",
       "      <td>ENSMUSG00000063889.16</td>\n",
       "      <td>Crem</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>99.960064</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000144512</th>\n",
       "      <td>ENCODEMT000144512</td>\n",
       "      <td>ENSMUSG00000040374.13</td>\n",
       "      <td>Pex2</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>99.960064</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000145637</th>\n",
       "      <td>ENCODEMT000145637</td>\n",
       "      <td>ENSMUSG00000059734.7</td>\n",
       "      <td>Ndufs8</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>99.960064</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000147256</th>\n",
       "      <td>ENCODEMT000147256</td>\n",
       "      <td>ENSMUSG00000045573.9</td>\n",
       "      <td>Penk</td>\n",
       "      <td>NNC</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001997</td>\n",
       "      <td>99.840256</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCODEMT000151946</th>\n",
       "      <td>ENCODEMT000151946</td>\n",
       "      <td>ENSMUSG00000096255.2</td>\n",
       "      <td>Dynlt1b</td>\n",
       "      <td>NIC</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>99.960064</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  annot_transcript_name          annot_gene_id  \\\n",
       "tid                                                              \n",
       "ENCODEMT000142671     ENCODEMT000142671  ENSMUSG00000063889.16   \n",
       "ENCODEMT000144512     ENCODEMT000144512  ENSMUSG00000040374.13   \n",
       "ENCODEMT000145637     ENCODEMT000145637   ENSMUSG00000059734.7   \n",
       "ENCODEMT000147256     ENCODEMT000147256   ENSMUSG00000045573.9   \n",
       "ENCODEMT000151946     ENCODEMT000151946   ENSMUSG00000096255.2   \n",
       "\n",
       "                  annot_gene_name transcript_novelty     mt  \\\n",
       "tid                                                           \n",
       "ENCODEMT000142671            Crem                NIC  False   \n",
       "ENCODEMT000144512            Pex2                NIC  False   \n",
       "ENCODEMT000145637          Ndufs8                NIC  False   \n",
       "ENCODEMT000147256            Penk                NNC  False   \n",
       "ENCODEMT000151946         Dynlt1b                NIC  False   \n",
       "\n",
       "                   n_cells_by_counts  mean_counts  pct_dropout_by_counts  \\\n",
       "tid                                                                        \n",
       "ENCODEMT000142671                  1     0.000399              99.960064   \n",
       "ENCODEMT000144512                  1     0.000399              99.960064   \n",
       "ENCODEMT000145637                  1     0.000399              99.960064   \n",
       "ENCODEMT000147256                  4     0.001997              99.840256   \n",
       "ENCODEMT000151946                  1     0.000399              99.960064   \n",
       "\n",
       "                   total_counts  \n",
       "tid                              \n",
       "ENCODEMT000142671           1.0  \n",
       "ENCODEMT000144512           1.0  \n",
       "ENCODEMT000145637           1.0  \n",
       "ENCODEMT000147256           5.0  \n",
       "ENCODEMT000151946           1.0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.reset_index(inplace=True)\n",
    "adata.var.set_index('annot_transcript_id', inplace=True)\n",
    "adata.var.index.name = 'tid'\n",
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "developmental-premiere",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc</th>\n",
       "      <th>bc3_long</th>\n",
       "      <th>bc2_long</th>\n",
       "      <th>bc1_long</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sr_umi_count</th>\n",
       "      <th>sr_gene_count</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>cellID</th>\n",
       "      <th>sr_clusters</th>\n",
       "      <th>...</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>leiden</th>\n",
       "      <th>lr_celltype</th>\n",
       "      <th>umap_x</th>\n",
       "      <th>umap_y</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATCGAACCGAGACTGTCCCG-ont_b</th>\n",
       "      <td>AAACATCGAACCGAGACTGTCCCG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AACCGAGA</td>\n",
       "      <td>CTGTCCCG</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>61.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.196721</td>\n",
       "      <td>L9</td>\n",
       "      <td>X-Zone</td>\n",
       "      <td>8.509377</td>\n",
       "      <td>8.825970</td>\n",
       "      <td>28</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAAGACGGAAGCGAAAC-ont_b</th>\n",
       "      <td>AAACATCGAAGACGGAAGCGAAAC</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AAGACGGA</td>\n",
       "      <td>AGCGAAAC</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>91.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>L1</td>\n",
       "      <td>Cortex_ZF</td>\n",
       "      <td>13.440162</td>\n",
       "      <td>5.323537</td>\n",
       "      <td>40</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAAGGACACCTGTCCCG-pb_2ka</th>\n",
       "      <td>AAACATCGAAGGACACCTGTCCCG</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AAGGACAC</td>\n",
       "      <td>CTGTCCCG</td>\n",
       "      <td>pb_2ka</td>\n",
       "      <td>10119.0</td>\n",
       "      <td>4010.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>AAACATCGAAGGACAC_14.ENCLB593HHL</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>83.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.409638</td>\n",
       "      <td>L5</td>\n",
       "      <td>Medulla_EPI</td>\n",
       "      <td>4.639245</td>\n",
       "      <td>17.013636</td>\n",
       "      <td>53</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGAATCCGTCAGCGAAAC-ont_b</th>\n",
       "      <td>AAACATCGAATCCGTCAGCGAAAC</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>AATCCGTC</td>\n",
       "      <td>AGCGAAAC</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>126.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.174603</td>\n",
       "      <td>L2</td>\n",
       "      <td>Adipocytes</td>\n",
       "      <td>5.295306</td>\n",
       "      <td>11.376595</td>\n",
       "      <td>60</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGACGTATCAGTTACGTA-ont_b</th>\n",
       "      <td>AAACATCGACGTATCAGTTACGTA</td>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>ACGTATCA</td>\n",
       "      <td>GTTACGTA</td>\n",
       "      <td>ont_b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>114.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.017544</td>\n",
       "      <td>L0</td>\n",
       "      <td>Cortex_ZG</td>\n",
       "      <td>13.883456</td>\n",
       "      <td>6.713534</td>\n",
       "      <td>33</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       bc  bc3_long  bc2_long  \\\n",
       "dataset                                                                         \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b   AAACATCGAACCGAGACTGTCCCG  AAACATCG  AACCGAGA   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b   AAACATCGAAGACGGAAGCGAAAC  AAACATCG  AAGACGGA   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka  AAACATCGAAGGACACCTGTCCCG  AAACATCG  AAGGACAC   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b   AAACATCGAATCCGTCAGCGAAAC  AAACATCG  AATCCGTC   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b   AAACATCGACGTATCAGTTACGTA  AAACATCG  ACGTATCA   \n",
       "\n",
       "                                 bc1_long experiment  sr_umi_count  \\\n",
       "dataset                                                              \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b   CTGTCCCG      ont_b           NaN   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b   AGCGAAAC      ont_b           NaN   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka  CTGTCCCG     pb_2ka       10119.0   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b   AGCGAAAC      ont_b           NaN   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b   GTTACGTA      ont_b           NaN   \n",
       "\n",
       "                                 sr_gene_count  Unnamed: 0  \\\n",
       "dataset                                                      \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b             NaN         NaN   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b             NaN         NaN   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka         4010.0       301.0   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b             NaN         NaN   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b             NaN         NaN   \n",
       "\n",
       "                                                          cellID sr_clusters  \\\n",
       "dataset                                                                        \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b                               NaN         NaN   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b                               NaN         NaN   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka  AAACATCGAAGGACAC_14.ENCLB593HHL         3.0   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b                               NaN         NaN   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b                               NaN         NaN   \n",
       "\n",
       "                                 ... n_genes_by_counts total_counts  \\\n",
       "dataset                          ...                                  \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b   ...                28         61.0   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b   ...                40         91.0   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka  ...                53         83.0   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b   ...                60        126.0   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b   ...                33        114.0   \n",
       "\n",
       "                                total_counts_mt pct_counts_mt leiden  \\\n",
       "dataset                                                                \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b              5.0      8.196721     L9   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b              0.0      0.000000     L1   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka             2.0      2.409638     L5   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b              4.0      3.174603     L2   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b              8.0      7.017544     L0   \n",
       "\n",
       "                                 lr_celltype     umap_x     umap_y  n_genes  \\\n",
       "dataset                                                                       \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b        X-Zone   8.509377   8.825970       28   \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b     Cortex_ZF  13.440162   5.323537       40   \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka  Medulla_EPI   4.639245  17.013636       53   \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b    Adipocytes   5.295306  11.376595       60   \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b     Cortex_ZG  13.883456   6.713534       33   \n",
       "\n",
       "                                 n_counts  \n",
       "dataset                                    \n",
       "AAACATCGAACCGAGACTGTCCCG-ont_b       61.0  \n",
       "AAACATCGAAGACGGAAGCGAAAC-ont_b       91.0  \n",
       "AAACATCGAAGGACACCTGTCCCG-pb_2ka      83.0  \n",
       "AAACATCGAATCCGTCAGCGAAAC-ont_b      126.0  \n",
       "AAACATCGACGTATCAGTTACGTA-ont_b      114.0  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.index.name = 'dataset'\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "instructional-third",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from test_mousewg.p\n",
      "\n",
      "Adding abundance for datasets AAACATCGAACCGAGACTGTCCCG-ont_b, AAACATCGAAGACGGAAGCGAAAC-ont_b, AAACATCGAAGGACACCTGTCCCG-pb_2ka, AAACATCGAATCCGTCAGCGAAAC-ont_b, AAACATCGACGTATCAGTTACGTA-ont_b... (and 2413 more) to SwanGraph\n",
      "calculating tpm\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Shape of passed values is (2418, 1), indices imply (2418, 22262)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   1661\u001b[0m                 blocks = [\n\u001b[0;32m-> 1662\u001b[0;31m                     \u001b[0mmake_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mslice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1663\u001b[0m                 ]\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mmake_block\u001b[0;34m(values, placement, klass, ndim, dtype)\u001b[0m\n\u001b[1;32m   2718\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2719\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mndim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mplacement\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2720\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, values, placement, ndim)\u001b[0m\n\u001b[1;32m   2374\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2375\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mndim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mplacement\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2376\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, values, placement, ndim)\u001b[0m\n\u001b[1;32m    129\u001b[0m             raise ValueError(\n\u001b[0;32m--> 130\u001b[0;31m                 \u001b[0;34mf\"Wrong number of items passed {len(self.values)}, \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    131\u001b[0m                 \u001b[0;34mf\"placement implies {len(self.mgr_locs)}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Wrong number of items passed 1, placement implies 22262",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-52-b5f78d900788>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'test_mousewg.p'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0madd_h5ad\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-51-6fab95ddbf14>\u001b[0m in \u001b[0;36madd_h5ad\u001b[0;34m(sg, adata)\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[0;31m# HAVE to use outer join for concatenating adatas though!\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m     \u001b[0mconcat_adata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-49-7dc9fe83e43e>\u001b[0m in \u001b[0;36mconcat_adata\u001b[0;34m(sg, adata)\u001b[0m\n\u001b[1;32m     34\u001b[0m         \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'counts'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'calculating tpm'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m         \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tpm'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcalc_tpm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mt_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0;31m# could probably parallelize calc_pi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py\u001b[0m in \u001b[0;36mcalc_tpm\u001b[0;34m(adata, sg_df, obs_col)\u001b[0m\n\u001b[1;32m    439\u001b[0m         \u001b[0mid_col\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    440\u001b[0m         \u001b[0mconditions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mobs_col\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 441\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcalc_total_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobs_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobs_col\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    442\u001b[0m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/swan_vis/utils.py\u001b[0m in \u001b[0;36mcalc_total_counts\u001b[0;34m(adata, obs_col, layer)\u001b[0m\n\u001b[1;32m    329\u001b[0m         \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlayers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlayer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m \tdf = pd.DataFrame(data=adata.X, index=adata.obs[obs_col].tolist(), \\\n\u001b[0;32m--> 331\u001b[0;31m \t\tcolumns=adata.var.index.tolist())\n\u001b[0m\u001b[1;32m    332\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    333\u001b[0m         \u001b[0;31m# add up values on condition (row)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    521\u001b[0m                     \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marrays_to_mgr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    522\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m                     \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_ndarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    525\u001b[0m                 \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36minit_ndarray\u001b[0;34m(values, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    232\u001b[0m         \u001b[0mblock_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    233\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 234\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mcreate_block_manager_from_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblock_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    235\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    236\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   1670\u001b[0m         \u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"values\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1671\u001b[0m         \u001b[0mtot_items\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1672\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mconstruction_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtot_items\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1673\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1674\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (2418, 1), indices imply (2418, 22262)"
     ]
    }
   ],
   "source": [
    "sg = swan.read('test_mousewg.p')\n",
    "add_h5ad(sg, adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "purple-mainland",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_h5ad(sg, adata):\n",
    "    \"\"\"\n",
    "    Adds abundance and metadata from an h5ad file to the SwanGraph. Transcripts\n",
    "    in the SwanGraph but not in the h5ad file will be assigned 0 counts.\n",
    "    Transcripts in the h5ad but not in the SwanGraph will not be added.\n",
    "    \n",
    "    Parameters:\n",
    "        fname (str): Path to h5ad file.\n",
    "            Var table should have the transcript id ('tid') as the index\n",
    "            Obs table should have the dataset ('dataset') as the index\n",
    "    \"\"\"\n",
    "    \n",
    "#     # read in h5ad file\n",
    "#     swan.check_file_loc(fname, 'h5ad')\n",
    "#     try:\n",
    "#         adata = sc.read(fname)\n",
    "#     except:\n",
    "#         raise ValueError('Problem reading h5ad {}'.format(counts_file))\n",
    "        \n",
    "    # limit to just the transcripts already in the graph\n",
    "    sg_tids = sg.t_df.tid.tolist()\n",
    "    ab_tids = adata.var.index.tolist()\n",
    "    tids = list(set(sg_tids)&(set(ab_tids)))\n",
    "\n",
    "    # remove entries in the adata that aren't already in the SwanGraph\n",
    "    adata = adata[:, tids]\n",
    "    \n",
    "    # double index obs and var\n",
    "    adata.obs['dataset'] = adata.obs.index.tolist()\n",
    "    adata.var['tid'] = adata.var.index.tolist()\n",
    "    \n",
    "    # for transcripts that aren't already in the adata, add them as 0 counts \n",
    "    # Or actually maybe not for now\n",
    "    # HAVE to use outer join for concatenating adatas though! \n",
    "    \n",
    "    concat_adata(sg, adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cathedral-broad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_adata(sg, adata):\n",
    "    \"\"\"\n",
    "    Once we have an anndata object (either generated from a flat counts file\n",
    "    or from an h5ad file), add the abundance from it to the SwanGraph.\n",
    "    \n",
    "    Parameters:\n",
    "        adata (anndata AnnData): AnnData from add_abundance or add_h5ad.\n",
    "    \"\"\"\n",
    "    \n",
    "    # add each dataset to list of \"datasets\", check if any are already there!\n",
    "    datasets = adata.obs.index.tolist()\n",
    "    for d in datasets:\n",
    "        if d in sg.datasets:\n",
    "            raise ValueError('Dataset {} already present in the SwanGraph.'.format(d))\n",
    "    sg.datasets.extend(datasets)\n",
    "    \n",
    "    print()\n",
    "    if len(datasets) <= 5:\n",
    "        print('Adding abundance for datasets {} to SwanGraph.'.format(', '.join(datasets)))\n",
    "    else:\n",
    "        mini_datasets = datasets[:5]\n",
    "        n = len(datasets) - len(mini_datasets)\n",
    "        print('Adding abundance for datasets {}... (and {} more) to SwanGraph'.format(', '.join(mini_datasets), n))\n",
    "    \n",
    "    \n",
    "    # if there is preexisting abundance data in the SwanGraph, concatenate\n",
    "    # otherwise, adata is the new transcript level adata\n",
    "    if not sg.has_abundance():\n",
    "\n",
    "        # create transcript-level adata object\n",
    "        sg.adata = adata\n",
    "\n",
    "        # add counts as layers\n",
    "        sg.adata.layers['counts'] = sg.adata.X\n",
    "        print('calculating tpm')\n",
    "        sg.adata.layers['tpm'] = swan.calc_tpm(sg.adata, sg.t_df).to_numpy()\n",
    "\n",
    "        # could probably parallelize calc_pi\n",
    "        if not sg.sc:\n",
    "            sg.adata.layers['pi'] = swan.calc_pi(sg.adata, sg.t_df)[0].to_numpy()\n",
    "    else:\n",
    "        sg.adata = sg.adata.concat(adata, join='outer')\n",
    "#         # concatenate the raw, tpm, and pi data\n",
    "#         X = np.concatenate((sg.adata.layers['counts'], X), axis=0)\n",
    "\n",
    "#         # concatenate obs\n",
    "#         obs = pd.concat([sg.adata.obs, obs], axis=0)\n",
    "\n",
    "#         # construct a new adata from the concatenated objects\n",
    "#         adata = anndata.AnnData(var=var, obs=obs, X=X)\n",
    "#         adata.layers['counts'] = X\n",
    "\n",
    "#         # some cleanup for unstructured data\n",
    "#         adata.uns = sg.adata.uns\n",
    "\n",
    "#         # set anndata to new guy\n",
    "#         sg.adata = adata\n",
    "\n",
    "        # recalculate pi and tpm\n",
    "        sg.adata.layers['tpm'] = swan.calc_tpm(sg.adata, sg.t_df).to_numpy()\n",
    "\n",
    "        if not sg.sc:\n",
    "            sg.adata.layers['pi'] = swan.calc_pi(sg.adata, sg.t_df)[0].to_numpy()\n",
    "\n",
    "\n",
    "    # add abundance for edges, TSS per gene, and TES per gene\n",
    "    # print('adding edge abundance')\n",
    "    # sg.create_edge_adata()\n",
    "    print('adding tss abundance')\n",
    "    swan.create_end_adata(sg, kind='tss')\n",
    "    print('adding tes abundance')\n",
    "    swan.create_end_adata(sg, kind='tes')\n",
    "\n",
    "    # set abundance flag to true\n",
    "    sg.abundance = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "rough-swaziland",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_abundance(self, counts_file):\n",
    "    \"\"\"\n",
    "    Adds abundance from a counts matrix to the SwanGraph. Transcripts in the\n",
    "    SwanGraph but not in the counts matrix will be assigned 0 counts.\n",
    "    Transcripts in the abundance matrix but not in the SwanGraph will not\n",
    "    have expression added.\n",
    "\n",
    "    Parameters:\n",
    "        counts_file (str): Path to TSV expression file where first column is\n",
    "            the transcript ID and following columns name the added datasets and\n",
    "            their counts in each dataset, OR to a TALON abundance matrix.\n",
    "    \"\"\"\n",
    "\n",
    "    # read in abundance file\n",
    "    check_file_loc(counts_file, 'abundance matrix')\n",
    "    try:\n",
    "        df = pd.read_csv(counts_file, sep='\\t')\n",
    "    except:\n",
    "        raise ValueError('Problem reading expression matrix {}'.format(counts_file))\n",
    "\n",
    "    # check if abundance matrix is a talon abundance matrix\n",
    "    cols = ['gene_ID', 'transcript_ID', 'annot_gene_id', 'annot_transcript_id',\n",
    "        'annot_gene_name', 'annot_transcript_name', 'n_exons', 'length',\n",
    "        'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
    "    if df.columns.tolist()[:11] == cols:\n",
    "        df = reformat_talon_abundance(counts_file)\n",
    "\n",
    "    # rename transcript ID column\n",
    "    col = df.columns[0]\n",
    "    df.rename({col: 'tid'}, axis=1, inplace=True)\n",
    "\n",
    "    # limit to just the transcripts already in the graph\n",
    "    sg_tids = sg.t_df.tid.tolist()\n",
    "    ab_tids = df.tid.tolist()\n",
    "\n",
    "    # right merge to keep everything in the t_df already\n",
    "    df = df.merge(sg.t_df['tid'].to_frame(), how='right', left_on='tid', right_index=True)\n",
    "    df.drop(['tid_x', 'tid_y'], axis=1, inplace=True)\n",
    "\n",
    "    # fill NaNs with 0 counts\n",
    "    df.fillna(0, inplace=True)\n",
    "    df.set_index('tid', inplace=True)\n",
    "\n",
    "    # transpose to get adata format\n",
    "    df = df.T\n",
    "\n",
    "    # get adata components - obs, var, and X\n",
    "    var = df.columns.to_frame()\n",
    "    var.columns = ['tid']\n",
    "    obs = df.index.to_frame()\n",
    "    obs.columns = ['dataset']\n",
    "    X = df.to_numpy()\n",
    "\n",
    "    adata = anndata.AnnData(var=var, obs=obs, X=X)\n",
    "    \n",
    "    sg.concat_adata(sg, adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sexual-attitude",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
