{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "spoken-campbell",
   "metadata": {},
   "outputs": [],
   "source": [
    "import swan_vis as swan\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import scipy\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fuzzy-turning",
   "metadata": {},
   "outputs": [],
   "source": [
    "# utils.py\n",
    "def calc_total_counts(adata, obs_col='dataset', layer='counts'):\n",
    "\t\"\"\"\n",
    "\tCalculate cumulative expression per adata entry based on condition given\n",
    "\tby `obs_col`. Default column to use is `adata.obs` index column, `dataset`.\n",
    "\n",
    "\tParameters:\n",
    "\t\tadata (anndata AnnData): Annotated data object from the SwanGraph\n",
    "\t\tobs_col (str): Column name from adata.obs table to group on.\n",
    "\t\t\tDefault: 'dataset'\n",
    "\t\tlayer (str): Layer of AnnData to pull from. Default = 'counts'\n",
    "\n",
    "\tReturns:\n",
    "\t\tdf (pandas DataFrame): Pandas DataFrame where rows are the different\n",
    "\t\t\tconditions from `obs_col` and the columns are transcript ids in the\n",
    "\t\t\tSwanGraph, and values represent the cumulative counts per isoform\n",
    "\t\t\tper condition.\n",
    "\n",
    "\t\"\"\"\n",
    "\tadata.X = adata.layers[layer]\n",
    "\tdf = pd.DataFrame(data=adata.X, index=adata.obs[obs_col].tolist(), \\\n",
    "\t\tcolumns=adata.var.index.tolist())\n",
    "\n",
    "\t# add up values on condition (row)\n",
    "\tdf = df.groupby(level=0).sum()\n",
    "\n",
    "\t# df = df.transpose()\n",
    "\n",
    "\treturn df\n",
    "\n",
    "def calc_tpm(adata, obs_col='dataset'):\n",
    "    \"\"\"\n",
    "\tCalculate the TPM per condition given by `obs_col`.\n",
    "\tDefault column to use is `adata.obs` index column, `dataset`.\n",
    "\n",
    "\tParameters:\n",
    "\t\tadata (anndata AnnData): Annotated data object from the SwanGraph\n",
    "\t\tobs_col (str or list of str): Column name from adata.obs table to group on.\n",
    "\t\t\tDefault: 'dataset'\n",
    "\n",
    "\tReturns:\n",
    "\t\tdf (pandas DataFrame): Pandas DataFrame where rows are the different\n",
    "\t\t\tconditions from `obs_col` and the columns are observed transcript ids in the\n",
    "\t\t\tSwanGraph, and values represent the TPM value per isoform per\n",
    "\t\t\tcondition.\n",
    "\t\"\"\"\n",
    "    \n",
    "    # first normalize across individual datasets / cells\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "# def calc_tpm(adata, sg_df=None, obs_col='dataset'):\n",
    "# \t\"\"\"\n",
    "# \tCalculate the TPM per condition given by `obs_col`.\n",
    "# \tDefault column to use is `adata.obs` index column, `dataset`.\n",
    "\n",
    "# \tParameters:\n",
    "# \t\tadata (anndata AnnData): Annotated data object from the SwanGraph\n",
    "# \t\tsg_df (pandas DataFrame): Pandas DataFrame from SwanGraph that will\n",
    "# \t\t\tbe used to order the rows of resultant TPM DataFrame\n",
    "# \t\tobs_col (str or list of str): Column name from adata.obs table to group on.\n",
    "# \t\t\tDefault: 'dataset'\n",
    "\n",
    "# \tReturns:\n",
    "# \t\tdf (pandas DataFrame): Pandas DataFrame where rows are the different\n",
    "# \t\t\tconditions from `obs_col` and the columns are observed transcript ids in the\n",
    "# \t\t\tSwanGraph, and values represent the TPM value per isoform per\n",
    "# \t\t\tcondition.\n",
    "# \t\"\"\"\n",
    "\n",
    "# \t# calculate cumulative counts across obs_col\n",
    "# \tid_col = adata.var.index.name\n",
    "# \tconditions = adata.obs[obs_col].unique().tolist()\n",
    "# \tdf = swan.calc_total_counts(adata, obs_col=obs_col)\n",
    "# \tdf = df.transpose()\n",
    "    \n",
    "# \tprint(df.shape)\n",
    "\n",
    "# \t# we use ints to index edges and locs\n",
    "# \tif id_col == 'vertex_id' or id_col == 'edge_id':\n",
    "# \t\tdf.index = df.index.astype('int')\n",
    "\n",
    "# \t# calculate tpm per isoform per condition\n",
    "# \ttpm_cols = []\n",
    "# \tfor c in conditions:\n",
    "# \t\tcond_col = '{}_tpm'.format(c)\n",
    "# \t\ttotal_col = '{}_total'.format(c)\n",
    "# \t\tdf[total_col] = df[c].sum()\n",
    "# \t\tdf[cond_col] = (df[c]*1000000)/df[total_col]\n",
    "# \t\ttpm_cols.append(cond_col)\n",
    "\n",
    "# \t# formatting\n",
    "# \tdf.index.name = id_col\n",
    "# \tdf = df[tpm_cols]\n",
    "# \tfor col in tpm_cols:\n",
    "# \t\tnew_col = col[:-4]\n",
    "# \t\tdf.rename({col: new_col}, axis=1, inplace=True)\n",
    "\n",
    "# \t# reorder columns like adata.obs\n",
    "# \tdf = df[adata.obs[obs_col].unique().tolist()]\n",
    "# \tdf = df.transpose()\n",
    "\n",
    "# \t# reorder in adata.var / t_df order\n",
    "# \tif not isinstance(sg_df, type(None)):\n",
    "# \t\tids = [tid for tid in sg_df[id_col].tolist() if tid in df.columns.tolist()]\n",
    "# \t\tdf = df[sg_df[id_col].tolist()]\n",
    "\n",
    "# \treturn df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "unlike-psychology",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_abundance(sg, counts_file):\n",
    "    \"\"\"\n",
    "    Adds abundance from a counts matrix to the SwanGraph. Transcripts in the\n",
    "    SwanGraph but not in the counts matrix will be assigned 0 counts.\n",
    "    Transcripts in the abundance matrix but not in the SwanGraph will not\n",
    "    have expression added.\n",
    "\n",
    "    Parameters:\n",
    "        counts_file (str): Path to TSV expression file where first column is\n",
    "            the transcript ID and following columns name the added datasets and\n",
    "            their counts in each dataset, OR to a TALON abundance matrix.\n",
    "    \"\"\"\n",
    "\n",
    "    # read in abundance file\n",
    "    swan.check_file_loc(counts_file, 'abundance matrix')\n",
    "    try:\n",
    "        df = pd.read_csv(counts_file, sep='\\t')\n",
    "    except:\n",
    "        raise ValueError('Problem reading expression matrix {}'.format(counts_file))\n",
    "\n",
    "    # check if abundance matrix is a talon abundance matrix\n",
    "    cols = ['gene_ID', 'transcript_ID', 'annot_gene_id', 'annot_transcript_id',\n",
    "        'annot_gene_name', 'annot_transcript_name', 'n_exons', 'length',\n",
    "        'gene_novelty', 'transcript_novelty', 'ISM_subtype']\n",
    "    if df.columns.tolist()[:11] == cols:\n",
    "        df = swan.reformat_talon_abundance(counts_file)\n",
    "\n",
    "    # rename transcript ID column\n",
    "    col = df.columns[0]\n",
    "    df.rename({col: 'tid'}, axis=1, inplace=True)\n",
    "\n",
    "    # limit to just the transcripts already in the graph\n",
    "    sg_tids = sg.t_df.tid.tolist()\n",
    "    ab_tids = df.tid.tolist()\n",
    "    tids = list(set(sg_tids)&set(ab_tids))\n",
    "    df = df.loc[df.tid.isin(tids)]\n",
    "    \n",
    "    # transpose to get adata format\n",
    "    df.set_index('tid', inplace=True)\n",
    "    df = df.T\n",
    "    \n",
    "    # get adata components - obs, var, and X\n",
    "    var = df.columns.to_frame()\n",
    "    var.columns = ['tid']\n",
    "    obs = df.index.to_frame()\n",
    "    obs.columns = ['dataset']\n",
    "    X = df.to_numpy()\n",
    "    \n",
    "    # create transcript-level adata object\n",
    "    adata = anndata.AnnData(var=var, obs=obs, X=X)\n",
    "\n",
    "    # add each dataset to list of \"datasets\", check if any are already there!\n",
    "    datasets = adata.obs.dataset.tolist()\n",
    "    for d in datasets:\n",
    "        if d in sg.datasets:\n",
    "            raise ValueError('Dataset {} already present in the SwanGraph.'.format(d))\n",
    "    sg.datasets.extend(datasets)\n",
    "\n",
    "    print()\n",
    "    if len(datasets) <= 5:\n",
    "        print('Adding abundance for datasets {} to SwanGraph.'.format(', '.join(datasets)))\n",
    "    else:\n",
    "        mini_datasets = datasets[:5]\n",
    "        n = len(datasets) - len(mini_datasets)\n",
    "        print('Adding abundance for datasets {}... (and {} more) to SwanGraph'.format(', '.join(mini_datasets), n))\n",
    "\n",
    "    # if there is preexisting abundance data in the SwanGraph, concatenate\n",
    "    # otherwise, adata is the new transcript level adata\n",
    "    if not sg.has_abundance():\n",
    "\n",
    "        # create transcript-level adata object\n",
    "        sg.adata = adata\n",
    "\n",
    "        # add counts as layers\n",
    "        sg.adata.layers['counts'] = sg.adata.X\n",
    "#         print('calculating tpm')\n",
    "#         sg.adata.layers['tpm'] = calc_tpm(sg.adata, sg.t_df).to_numpy()\n",
    "\n",
    "#         # could probably parallelize calc_pi\n",
    "#         if not sg.sc:\n",
    "#             print('calculating tpm') \n",
    "#             sg.adata.layers['pi'] = swan.calc_pi(sg.adata, sg.t_df)[0].to_numpy()\n",
    "    else:\n",
    "        \n",
    "        # first set current layer to be counts\n",
    "        sg.adata.X = sg.adata.layers['counts']\n",
    "        \n",
    "        # concatenate existing adata with new one\n",
    "        # outer join to add all new transcripts (that are from added\n",
    "        # annotation or transcriptome) to the abundance\n",
    "        uns = sg.adata.uns\n",
    "        sg.adata = sg.adata.concatenate(adata, join='outer', index_unique=None)\n",
    "        sg.adata.uns = uns\n",
    "        \n",
    "#         # recalculate pi and tpm\n",
    "#         print('calculating tpm')\n",
    "#         sg.adata.layers['tpm'] = calc_tpm(sg.adata, sg.t_df).to_numpy()\n",
    "\n",
    "#         if not sg.sc:\n",
    "#             print('calculating pi')\n",
    "#             sg.adata.layers['pi'] = swan.calc_pi(sg.adata, sg.t_df)[0].to_numpy()\n",
    "\n",
    "#     # add abundance for edges, TSS per gene, and TES per gene\n",
    "#     # sg.create_edge_adata()\n",
    "#     sg.create_end_adata(kind='tss')\n",
    "#     sg.create_end_adata(kind='tes')\n",
    "\n",
    "    # set abundance flag to true\n",
    "    sg.abundance = True\n",
    "    \n",
    "    return sg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "broke-jaguar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from test_mousewg.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read('test_mousewg.p')\n",
    "ab = '/Users/fairliereese/mortazavi_lab/data/mousewg/lr_bulk/talon/mouse_talon_abundance_filtered.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "figured-modification",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding abundance for datasets gastroc_14d_f_2, gastroc_14d_f_1, heart_18-20mo_m_1, heart_18-20mo_m_2, heart_18-20mo_f_1... (and 86 more) to SwanGraph\n"
     ]
    }
   ],
   "source": [
    "# test adding de novo\n",
    "sg = add_abundance(sg, ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "derived-motor",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_tpm(adata, obs_col='dataset'):\n",
    "    \n",
    "    # calculate tpm using scanpy\n",
    "    d = sc.pp.normalize_total(adata,\n",
    "                              layer='counts',\n",
    "                              target_sum=1e6,\n",
    "                              key_added='total_counts',\n",
    "                              inplace=False)\n",
    "    adata.obs['total_counts'] = d['norm_factor']\n",
    "    \n",
    "    # turn into a sparse dataframe\n",
    "    cols = adata.var.index.tolist()\n",
    "    inds = adata.obs[obs_col].tolist()\n",
    "    data = d['X']\n",
    "    data = scipy.sparse.csr_matrix(data)\n",
    "    df = pd.DataFrame.sparse.from_spmatrix(data, index=inds, columns=cols)\n",
    "    df.index.name = obs_col    \n",
    "\n",
    "    # average across tpm\n",
    "    if obs_col != 'dataset':\n",
    "        df.reset_index(inplace=True)\n",
    "        df = df.groupby(obs_col).mean()\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "moved-german",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding abundance for datasets dataset1, dataset2 to SwanGraph.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test1</th>\n",
       "      <th>test2</th>\n",
       "      <th>test3</th>\n",
       "      <th>test4</th>\n",
       "      <th>test5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dataset1</th>\n",
       "      <td>166666.671875</td>\n",
       "      <td>333333.34375</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>333333.34375</td>\n",
       "      <td>166666.671875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset2</th>\n",
       "      <td>166666.671875</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>333333.34375</td>\n",
       "      <td>333333.34375</td>\n",
       "      <td>166666.671875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  test1         test2         test3         test4  \\\n",
       "dataset                                                             \n",
       "dataset1  166666.671875  333333.34375       0.00000  333333.34375   \n",
       "dataset2  166666.671875       0.00000  333333.34375  333333.34375   \n",
       "\n",
       "                  test5  \n",
       "dataset                  \n",
       "dataset1  166666.671875  \n",
       "dataset2  166666.671875  "
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test_calc_tpm_1\n",
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../testing/files/test_full.gtf')\n",
    "sg = add_abundance(sg, '../testing/files/test_ab_1.tsv')\n",
    "sg.adata.obs['cluster'] = ['c1', 'c1']\n",
    "\n",
    "df = calc_tpm(sg.adata)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "wound-marshall",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding abundance for datasets dataset1, dataset2 to SwanGraph.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test1</th>\n",
       "      <th>test2</th>\n",
       "      <th>test3</th>\n",
       "      <th>test4</th>\n",
       "      <th>test5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>c1</th>\n",
       "      <td>166666.671875</td>\n",
       "      <td>166666.671875</td>\n",
       "      <td>166666.671875</td>\n",
       "      <td>333333.34375</td>\n",
       "      <td>166666.671875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 test1          test2          test3         test4  \\\n",
       "cluster                                                              \n",
       "c1       166666.671875  166666.671875  166666.671875  333333.34375   \n",
       "\n",
       "                 test5  \n",
       "cluster                 \n",
       "c1       166666.671875  "
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test_calc_tpm_2\n",
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../testing/files/test_full.gtf')\n",
    "sg = add_abundance(sg, '../testing/files/test_ab_1.tsv')\n",
    "sg.adata.obs['cluster'] = ['c1', 'c1']\n",
    "\n",
    "df = calc_tpm(sg.adata, obs_col='cluster')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "commercial-trademark",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_total_counts(adata, obs_col='dataset', layer='counts'):\n",
    "    \n",
    "    # turn into a sparse dataframe\n",
    "    cols = adata.var.index.tolist()\n",
    "    inds = adata.obs[obs_col].tolist()\n",
    "    data = adata.layers[layer]\n",
    "    data = scipy.sparse.csr_matrix(data)\n",
    "    df = pd.DataFrame.sparse.from_spmatrix(data, index=inds, columns=cols)\n",
    "    df.index.name = obs_col \n",
    "\n",
    "    # add up values on condition (row)\n",
    "    df = df.groupby(level=0).sum()\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "civic-tunisia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding transcriptome to the SwanGraph\n",
      "\n",
      "Adding abundance for datasets dataset1, dataset2 to SwanGraph.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test1</th>\n",
       "      <th>test2</th>\n",
       "      <th>test3</th>\n",
       "      <th>test4</th>\n",
       "      <th>test5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dataset1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset2</th>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          test1  test2  test3  test4  test5\n",
       "dataset                                    \n",
       "dataset1    5.0   10.0    0.0   10.0    5.0\n",
       "dataset2    5.0    0.0   10.0   10.0    5.0"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test_calc_total_counts_1\n",
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../testing/files/test_full.gtf')\n",
    "sg = add_abundance(sg, '../testing/files/test_ab_1.tsv')\n",
    "sg.adata.obs['cluster'] = ['c1', 'c1']\n",
    "\n",
    "df = calc_total_counts(sg.adata)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "technical-saint",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_calc_total_counts_2\n",
    "sg = swan.SwanGraph()\n",
    "sg.add_transcriptome('../testing/files/test_full.gtf')\n",
    "sg = add_abundance(sg, '../testing/files/test_ab_1.tsv')\n",
    "sg.adata.obs['cluster'] = ['c1', 'c1']\n",
    "\n",
    "df = calc_total_counts(sg.adata)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minute-homework",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test merging when incoming adata has new transcripts that old did not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defined-division",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test merging when incoming adata has duplicate dataset names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "medical-subdivision",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
